### PR TITLE
CDAP-14692 switch stores to use StructuredTable

### PR DESCRIPTION
**Pull request by albertshau**
Switch to use system tables instead of datasets.

### CODE REVIEW

File: wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java [hunk 1]
```diff
@@ -96,28 +84,37 @@ protected boolean shouldCopyHeader(@Nullable NamespacedId workspaceId) {
...
+      ConnectionStore store = ConnectionStore.get(context);
+      return store.get(connectionId);
+    }, ConnectionNotFoundException.class);
+  }
+
+  protected Connection getValidatedConnection(NamespacedId connectionId, ConnectionType expectedType) {
+    return TransactionRunners.run(getContext(), context -> {
+      ConnectionStore store = ConnectionStore.get(context);
+      return getValidatedConnection(store, connectionId, expectedType);
+    }, ConnectionNotFoundException.class);
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java]: should this also include `IOException` ?

Comment (CuriousVini) [reply wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java]: Also what would happen if run method throws a RuntimeException `BadRequestException`?

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSHandler.java [hunk 2]
```diff
@@ -325,9 +319,14 @@ public void loadObject(HttpServiceRequest request, HttpServiceResponder responde
         .setScope(scope)
         .setProperties(properties)
         .build();
-      ws.writeWorkspaceMeta(workspaceMeta);
+      if (blob.isDirectory()) {
+        throw new BadRequestException("Path specified is not a file.");
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSHandler.java]: should we add blobpath to error message? Having contextual info in error message would be better?

File: wrangler-service/src/main/java/co/cask/wrangler/service/schema/SchemaRegistryHandler.java [hunk 3]
```diff
@@ -148,19 +132,24 @@ public void upload(HttpServiceRequest request, HttpServiceResponder responder,
   public void upload(HttpServiceRequest request, HttpServiceResponder responder, @PathParam("context") String namespace,
                      @PathParam("id") String id) {
     respond(request, responder, namespace, () -> {
-      byte[] bytes = null;
+      byte[] bytes;
       ByteBuffer content = request.getContent();
       if (content != null && content.hasRemaining()) {
         bytes = new byte[content.remaining()];
         content.get(bytes);
+      } else {
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/schema/SchemaRegistryHandler.java:139]: why this change is needed?

File: wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java [hunk 4]
```diff
@@ -41,30 +49,98 @@
  * A schema entry contains the actual schema bytes and is uniquely identified by the schema it belongs to and a
  * version number. A schema entry is added to a schema and can also be removed. In addition, each schema keeps track
  * of the latest schema entry.
+ *
+ * Schema information is stored in two tables, one for the schema metadata and one for the schema entries.
```
Comment (CuriousVini) [inline wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java:53]: Nice!  +1 on javadoc

**Review by CuriousVini — COMMENTED** [review]
Should we add IOException as expected Exception type?

File: wrangler-storage/src/main/java/co/cask/wrangler/dataset/connections/ConnectionStore.java [hunk 5]
```diff
@@ -61,18 +70,44 @@
...
+  private static final String NAMESPACE_COL = "namespace";
+  private static final String ID_COL = "id";
   private static final String TYPE_COL = "type";
   private static final String NAME_COL = "name";
   private static final String DESC_COL = "description";
   private static final String PROPERTIES_COL = "properties";
   private static final String CREATED_COL = "created";
   private static final String UPDATED_COL = "updated";
-  private final Table table;
+  private static final StructuredTableId TABLE_ID = new StructuredTableId("connections");
```
Comment (CuriousVini) [inline wrangler-storage/src/main/java/co/cask/wrangler/dataset/connections/ConnectionStore.java:70]: should we also change javadoc for this store?

File: wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java [hunk 6]
```diff
@@ -41,30 +49,98 @@
...
+ * schema entry version, and the auto column is used for generating version numbers for any new schema entries.
+ *
+ * The schema entry table contains four columns:
+ *
+ * namespace, id, version, schema
+ *
+ * The namespace, id, and version columns form the primary key, while the schema column contains the actual schema.
  */
 public final class SchemaRegistry  {
   public static final String DATASET_NAME = "schemaRegistry";
```
Comment (CuriousVini) [inline wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java]: unused?

**Review by CuriousVini — COMMENTED** [review]
code changes lgtm 👍  just a few minor comments

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSHandler.java (hunk 2 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSHandler.java]: fixed

File: wrangler-service/src/main/java/co/cask/wrangler/service/schema/SchemaRegistryHandler.java (hunk 3 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/schema/SchemaRegistryHandler.java:139]: since bytes is now used in a closure, it needs to be effectively final

File: wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java (hunk 6 above)
Comment (albertshau) [reply wrangler-storage/src/main/java/co/cask/wrangler/dataset/schema/SchemaRegistry.java]: fixed

File: wrangler-storage/src/main/java/co/cask/wrangler/dataset/connections/ConnectionStore.java (hunk 5 above)
Comment (albertshau) [reply wrangler-storage/src/main/java/co/cask/wrangler/dataset/connections/ConnectionStore.java:70]: added a couple lines about the schema and primary key.

File: wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java (hunk 1 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerHandler.java]: hm, actually I don't need to explicitly mention any of these exceptions since they're all RuntimeException, which gets propagated by TransactionsRunners. Will change these.

**Comment by albertshau** [comment]
In terms of adding IOException, I don't think it will make a difference, as everything will get caught by the respond() method and return a 500.

**Review by CuriousVini — APPROVED** [review]
lgtm 👍
